fw/http_parser.c

```diff
@@ -3320,6 +3332,167 @@ __parse_pragma(TfwHttpMsg *hm, unsigned char *data, size_t len)
 }
 STACK_FRAME_NON_STANDARD(__parse_pragma);
 
+/**
+ * Parse Upgrade header field. Its semantics is described in RFC 7230 6.1.
+ * For now only websocket protocol supported.
+ *
+ * Nested FSM with explicit fine-grained fixups, should employ
+ * __FSM_I_MOVE_fixup()/__FSM_I_MATCH_fixup()/TRY_STR_fixup() everywhere.
+ */
+static int
+__req_parse_upgrade(TfwHttpMsg *hm, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st);
+
+	/*
+	 * Here we build a header value string manually to split it in chunks:
+	 * next chunk starts after ', ' list delimiter and '/' delimiter.
+	 * Optional protocol version chunk detached from protocol name chunk.
+	 */
+	__FSM_STATE(I_UpgradeProtocolStart) {
+		__FSM_I_MATCH_MOVE_fixup(token, I_UpgradeProtocol,
+					 TFW_STR_NAME);
+		/*
+		 * Protocol name should contain at least 1 character.
+		 */
+		if (likely(__fsm_sz))
+			__FSM_I_JMP(I_UpgradeProtocolEnd);
+
+		return CSTR_NEQ;
+	}
+
+	/*
+	 * At this state we know that we saw at least one character in
+	 * protocol name and now we can pass zero length token.
+	 */
+	__FSM_STATE(I_UpgradeProtocol) {
+		__FSM_I_MATCH_MOVE_fixup(token, I_UpgradeProtocol,
+					 TFW_STR_NAME);
+		__FSM_I_JMP(I_UpgradeProtocolEnd);
+	}
+
+	__FSM_STATE(I_UpgradeProtocolEnd) {
+		if (likely(__fsm_sz)) {
+			/* Save protocol name */
+			__msg_hdr_chunk_fixup(p, __fsm_sz);
+			__FSM_I_chunk_flags(TFW_STR_NAME);
+		}
```

Comment:
Usually we achieve this with some MOVE fixup macros, which moves the FSM and fixes up current string if we reach end of data

Reply:
The `__req_parse_upgrade()` function is low level and use manual chunking facilities. See comment before the function:

>  Nested FSM with explicit fine-grained fixups, should employ
 __FSM_I_MOVE_fixup()/__FSM_I_MATCH_fixup()/TRY_STR_fixup() everywhere.

All functions marked with this comments use the pattern you commented on here.